src/librustc_errors/emitter.rs

```diff
@@ -268,132 +269,216 @@ pub trait Emitter {
+                                                if macro_backtrace.len() > 2 && always_backtrace {
                                                     // only specify order when the macro
                                                     // backtrace is multiple levels deep
                                                     format!(" (#{})", i + 1)
                                                 } else {
                                                     String::new()
                                                 })));
-                    if !always_backtrace {
-                        break;
-                    }
+                }
+                if !always_backtrace {
+                    break;
                 }
             }
         }
+
         for (label_span, label_text) in new_labels {
             span.push_span_label(label_span, label_text);
         }
-        for sp_label in span.span_labels() {
-            if sp_label.span.is_dummy() {
-                continue;
-            }
-            if sm.span_to_filename(sp_label.span.clone()).is_macros() &&
-                !always_backtrace
-            {
-                let v = sp_label.span.macro_backtrace();
-                if let Some(use_site) = v.last() {
-                    before_after.push((sp_label.span.clone(), use_site.call_site.clone()));
-                }
-            }
+    }
+
+    // This does a small "fix" for multispans by looking to see if it can find any that
+    // point directly at <*macros>. Since these are often difficult to read, this
+    // will change the span to point at the use site.
+    fn fix_multispans_in_extern_macros(
+        &self,
+        source_map: &Option<Lrc<SourceMap>>,
+        span: &mut MultiSpan,
+        children: &mut Vec<SubDiagnostic>,
+    ) {
+        for span in iter::once(span).chain(children.iter_mut().map(|child| &mut child.span)) {
+            self.fix_multispan_in_extern_macros(source_map, span);
         }
+    }
+
+    // This "fixes" MultiSpans that contain Spans that are pointing to locations inside of
+    // <*macros>. Since these locations are often difficult to read, we move these Spans from
+    // <*macros> to their corresponding use site.
+    fn fix_multispan_in_extern_macros(
+        &self,
+        source_map: &Option<Lrc<SourceMap>>,
+        span: &mut MultiSpan,
+    ) {
+        let sm = match source_map {
+            Some(ref sm) => sm,
+            None => return,
+        };
+
+        // First, find all the spans in <*macros> and point instead at their use site
+        let replacements: Vec<(Span, Span)> =
+            span.primary_spans().iter().copied()
+                .chain(span.span_labels().iter().map(|sp_label| sp_label.span))
+                .filter_map(|sp| {
+                    if !sp.is_dummy() && sm.span_to_filename(sp).is_macros() {
+                        let maybe_callsite = sp.source_callsite();
```

Comment:
Why "maybe"?

Reply:
Because `sp.source_callsite()` will return `sp` if no macro invocation is actually involved.